File: CSharp/context_from_file/contextfromfile.cs
using System; // WHAT: Core .NET types like Exception and Console. WHY: We throw on missing keys and print output. DEPENDS ON: Step 1 (throw) + Step 4 (Console.WriteLine).
using System.Net.Http; // WHAT: HttpClient/HttpRequestMessage/HttpMethod. WHY: Required to send HTTP requests. DEPENDS ON: Step 3 (client + request).
using System.Net.Http.Headers; // WHAT: AuthenticationHeaderValue. WHY: Cleanly sets the Authorization header as "Bearer <key>". DEPENDS ON: Step 3 (req.Headers.Authorization).
using System.Text; // WHAT: Encoding.UTF8. WHY: Ensures JSON body is encoded as UTF-8. DEPENDS ON: Step 3 (StringContent).
using System.Text.Json; // WHAT: JsonSerializer + JsonDocument. WHY: Serialize request JSON + parse response JSON. DEPENDS ON: Step 3 (Serialize) + Step 4 (Parse/extract).
using System.Threading.Tasks; // WHAT: Async/await support for top-level statements. WHY: Allows awaiting HTTP calls in this file. DEPENDS ON: Step 4 (await SendAsync/ReadAsStringAsync).
using Microsoft.Extensions.Configuration; // WHAT: ConfigurationBuilder + IConfiguration. WHY: Reads User Secrets (OPENAI_API_KEY) safely. DEPENDS ON: Step 1 (config + apiKey).

// ============================================================
// Step 1: Load API key from environment (User Secrets) (Project 1 origin)
// ============================================================

// TASK: Build a configuration object that knows how to read values from this project's User Secrets store.
// WHY: This wires up the mechanism for fetching OPENAI_API_KEY later (without hardcoding it in source code).
// DEPENDS ON: `dotnet user-secrets init` (writes UserSecretsId into .csproj) + `dotnet user-secrets set "OPENAI_API_KEY" "..."`.
// SYNTAX: `new ConfigurationBuilder()` creates the builder; `.AddUserSecrets<Program>()` locates this project's secrets; `.Build()` returns IConfiguration.
var config = new ConfigurationBuilder()
  .AddUserSecrets<Program>()
  .Build();
// PYTHON: load_dotenv(override=True); api_key = os.getenv("OPENAI_API_KEY")
// R:      load_dot_env(file = file.path("..", ".env")); api_key <- Sys.getenv("OPENAI_API_KEY")

// TASK: Read the OpenAI API key into a variable for later HTTP headers.
// WHY: The key is required to authenticate requests (Authorization: Bearer ...).
// DEPENDS ON: `config` created above + User Secrets containing "OPENAI_API_KEY".
// SYNTAX: `config["OPENAI_API_KEY"]` reads a string; `?.Trim()` removes whitespace; `?? throw` fails fast if missing.
string apiKey = config["OPENAI_API_KEY"]?.Trim() ?? throw new Exception("OPENAI_API_KEY is missing");
// PYTHON: api_key = os.getenv("OPENAI_API_KEY")
// R:      api_key <- Sys.getenv("OPENAI_API_KEY")


// ============================================================
// Step 2: Define base URL / endpoint + auth header (Project 1 origin)
// ============================================================

// TASK: Create an HTTP client we can use to send requests.
// WHY: HttpClient is the standard way to send HTTP requests in .NET.
// DEPENDS ON: Nothing.
// SYNTAX: `new HttpClient()` constructs the client instance.
var client = new HttpClient();
// PYTHON: import requests
// R:      library(httr2)

// TASK: Store the Chat Completions endpoint URL in a variable.
// WHY: Keeping the URL in one place makes it easy to change later and avoids typos.
// DEPENDS ON: Nothing (this is a constant string).
// SYNTAX: A normal C# string variable assignment.
string baseUrl = "https://api.openai.com/v1/chat/completions";
// PYTHON: base_url = "https://api.openai.com/v1/chat/completions"
// R:      base_url <- "https://api.openai.com/v1/chat/completions"

// TASK: Build the Authorization header value using your API key.
// WHY: OpenAI expects `Authorization: Bearer <key>` on requests.
// DEPENDS ON: `apiKey` from Step 1.
// SYNTAX: String interpolation (`$"..."`) embeds apiKey into the header value.
string authHeader = $"Bearer {apiKey}";
// PYTHON: auth_header = f"Bearer {api_key}"
// R:      auth_header <- paste("Bearer", api_key)


// ============================================================
// Step 3: Build HTTP POST with JSON payload (Project 1 origin)
// ============================================================

// TASK: Build the JSON payload with model + messages.
// WHY: The API expects a JSON body with `model` and `messages` fields.
// DEPENDS ON: For now we send one hardcoded user message; later you’ll send the evolving `messages` history.
// SYNTAX: An anonymous object becomes JSON via `JsonSerializer.Serialize(payload)`.
var payload = new
{
  model = "gpt-4o-mini",
  messages = new object[]
  {
    new { role = "user", content = "Hello, how are you?" }
  }
};
// PYTHON: payload = {"model":"gpt-4o-mini","messages":[{"role":"user","content":"Hello, how are you?"}]}
// R:      payload <- list(model="gpt-4o-mini", messages=list(list(role="user", content="Hello, how are you?")))

// TASK: Create the POST request with headers + JSON body.
// WHY: This is the actual HTTP request we send to `baseUrl`.
// DEPENDS ON: `baseUrl` + `apiKey` from Steps 1–2 + `payload` from the block above.
// SYNTAX: `HttpRequestMessage` sets method+url; `Headers.Authorization` sets Bearer; `StringContent` sets JSON body with UTF-8 and content-type.
var req = new HttpRequestMessage(HttpMethod.Post, baseUrl);
req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
req.Content = new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");
// PYTHON: requests.post(base_url, headers={"Authorization": auth_header}, json=payload)
// R:      request(base_url) |> req_headers(Authorization=auth_header) |> req_body_json(payload) |> req_perform()

// ============================================================
// Step 4: Parse JSON response, extract assistant text (Project 1 origin)
// ============================================================

// TASK: Send the request and read the response body as text.
// WHY: We need the raw JSON string before we can parse it.
// DEPENDS ON: `client` + `req` created above.
// SYNTAX: `await` waits for the async HTTP call; `ReadAsStringAsync()` returns the body string.
var resp = await client.SendAsync(req);
string respText = await resp.Content.ReadAsStringAsync();
// PYTHON: resp = requests.post(...); resp_text = resp.text
// R:      response <- req_perform(...); resp_text <- resp_body_string(response)

// TASK: Parse JSON and extract `choices[0].message.content`.
// WHY: That field contains the assistant reply text.
// DEPENDS ON: `respText` containing valid JSON returned by the API.
// SYNTAX: `JsonDocument.Parse()` parses JSON; `GetProperty()` + `[0]` navigate to the nested value; `GetString()` returns the string.
using var doc = JsonDocument.Parse(respText);
string assistantText = doc.RootElement
  .GetProperty("choices")[0]
  .GetProperty("message")
  .GetProperty("content")
  .GetString() ?? "";
// PYTHON: assistant_text = resp.json()["choices"][0]["message"]["content"]
// R:      assistant_text <- resp_body_json(response)$choices[[1]]$message$content

// TASK: Print the assistant reply.
// WHY: This confirms the end-to-end request + parse pipeline works.
// DEPENDS ON: `assistantText` extracted above.
// SYNTAX: `Console.WriteLine()` writes to standard output.
Console.WriteLine(assistantText);
// PYTHON: print(assistant_text)
// R:      cat(assistant_text, "\n")

File: R/contextfromfile.R
library(dotenv) # Provides load_dot_env() so .env variables become visible to Sys.getenv()
library(httr2) # Modern HTTP client: build request, set JSON body, perform POST in a readable pipe

# If you always run from repo root: load_dot_env(file = ".env")
# If you run from R/: point at the parent folder's .env

# --- Step 1: Load API key from environment (Project 1) ---
load_dot_env(file = file.path("..", ".env")) # Loads OPENAI_API_KEY from project root without hardcoding the key
api_key <- Sys.getenv("OPENAI_API_KEY") # Reads the key into api_key for Authorization headers

# --- Step 2: Define base_url and headers (Project 1) ---
base_url <- "https://api.openai.com/v1/chat/completions" # Chat Completions API route; change here if you ever switch endpoints
auth_header <- paste("Bearer", api_key) # Full Authorization header OpenAI expects (Bearer + space + key from Step 1)

# --- Step 3: Build HTTP POST with JSON payload (Project 1) ---
call_openai <- function(messages) {
  request(base_url) |>
    req_headers(
      Authorization = auth_header, # Step 2: proves who we are to OpenAI
      `Content-Type` = "application/json" # Explicit JSON; matches what req_body_json sends
    ) |>
    req_body_json(list(
      model = "gpt-4o-mini", # Cheap, capable chat model; change if your org uses another name
      messages = messages # Full conversation history (system/user/assistant) as the API expects
    )) |>
    req_perform() # Sends the POST and returns the raw response object for Step 4 parsing
}

# --- Step 4: Parse JSON response, extract assistant text (Project 1) ---
extract_assistant_text <- function(response) {
  data <- resp_body_json(response) # Decode JSON body to an R list
  data$choices[[1]]$message$content # First completion's assistant message text
}

# --- Step 5: Define system_prompt string (Project 2) ---
system_prompt <- paste0(
  "You are my software engineering instructor. ",
  "You will help me learn how to program in any language.",
  "You will help me understand the concepts of programming. Lets get started by you explaining the syntax in human readable terms.",
  "Please explain the syntax in a way that is easy to understand and follow.",
  "Ask if I understand at each step and if not, explain again in a way that is easy to understand and follow."
)

# --- Step 6: Seed messages with system role (Project 2) ---
# CONCEPT: The transcript starts with the rules — messages is what you send every turn;
# role = "system" makes the model treat system_prompt as global policy before any user text.
messages <- list(list(role = "system", content = system_prompt)) # Initial transcript: policy first, before any user input

# --- Step 7: Build conversation loop with exit condition (Project 3) ---
# CONCEPT: repeat { ... } runs many turns until break (EOF or "quit").
repeat {
  # --- Step 8: Read user input from stdin (Project 3) ---
  # CONCEPT: stdin is keyboard when interactive, or piped lines when testing; warn=FALSE reduces EOF noise;
  # encoding="UTF-8" helps on Windows with non-ASCII text.
  cat("\nYou: ") # Prompt so it is obvious the program is waiting for input

  user_input <- readLines(
    con = "stdin", # Standard input from the terminal
    n = 1, # One line per turn
    warn = FALSE, # Fewer warnings when the stream ends (EOF) without a final newline
    encoding = "UTF-8" # Prefer UTF-8 so pasted/special characters match your editor and API text
  )

  if (length(user_input) == 0) {
    break # EOF (e.g. Ctrl+Z then Enter on Windows): exit before subscripting user_input
  }

  user_line <- trimws(user_input[[1]]) # Safe scalar string for quit check and for appending to messages

  if (tolower(user_line) == "quit") {
    break # Explicit exit word
  }

  # --- Step 9: Append user message to messages before the API call (Project 3) ---
  # CONCEPT: The API only sees what is in messages; append preserves full history.
  messages <- append(
    messages,
    list(list(role = "user", content = user_line)) # Record this turn's user text before the POST
  )

  # --- Step 10: Call API and append assistant reply (Project 3) ---
  # CONCEPT: Send full history, then store the assistant reply so the next turn has context (no amnesia).
  response <- call_openai(messages) # Full transcript (system + user/assistant turns) in one POST
  assistant_text <- extract_assistant_text(response) # Plain string from choices[[1]]$message$content
  cat("\nAssistant:", assistant_text, "\n") # Echo reply to the terminal

  messages <- append(
    messages,
    list(list(role = "assistant", content = assistant_text)) # Persist reply for the next user turn
  )
}

# --- Steps 11–12 (Project 4 — next): Read context.txt with safe path + UTF-8; inject into system_prompt
# before Step 6's messages <- ... (or move messages seeding below file load + injection).


# Step 11 — Read context.txt from disk (safe path + UTF-8) (Project 4, new)
# Build a path to context.txt next to your project (e.g. with file.path("..", "context.txt") when you run from R/), so it doesn’t depend on the current working directory in a fragile way.
# Read the file as UTF-8 and turn it into one string (e.g. paste(readLines(..., warn = FALSE), collapse = "\n"), with encoding handled as you prefer for your R version).
# Store that string in something like context_text (or a name you keep consistent with your README).

context_path <- file.path("..", "context.txt")   # # Stable path from R/ to project root; avoids hardcoding a drive letter
context_lines <- readLines(context_path, warn = FALSE, encoding = "UTF-8") #  # One string per line; UTF-8 for Windows/editor consistency
context_text <- paste(context_lines, collapse = "\n") # # Single block of text to inject into the system prompt in Step 12

# Rscript -e "p <- file.path('..','context.txt'); x <- paste(readLines(p, warn=FALSE, encoding='UTF-8'), collapse='\n'); cat(nchar(x) > 0, '\n')"

# Step 13: Inject file contents into system_prompt BEFORE seeding messages 
# (Project 4 origin)
# # ========================================= 
# CONCEPT: messages captures whatever system_prompt is at the moment you create it.
# # If you seed messages first, the model never sees your file 
# context.
# # So you must: 
# read context_text → build final system_prompt → then create messages.
# # In your chatbot.R, 
# this Step 13 code must appear right before the line
#  that seeds messages (your current Step 6).

#TASK: Save a copy of the original instructions before 
#adding the file context.
#WHY: We're about to overwrite system_prompt; without a backup we lose the base persona for any
#later rebuild
#DEPENDS ON: 'system_prompt'  from step 5 (Persona Bot)
#SYNTAX: Use '<-' to assign the current string into a new variable name. 
base_system_prompt <- system_prompt
# Python: base_system_prompt = system_prompt
# C# : string baseSystemPrompt = systemPrompt;


# Task: Create a system prompt by adding the context file text underneath the original instructions.
# WHY: The model only 'knows' what is in its system prompt -- injecting the file is how external context becomes useable. 
# DEPENDS ON: 'base_system_prompt' from line above, + 'context_text' from step 12 (file read)
# SYNTAX: paste0() concactenates strings with no separator; "\n\n" inserts a blank line for readability.

system_prompt <- paste0(base_system_prompt, "\n\n", context_text)
# Python: system_prompt = base_system_prompt + "\n\n" + context_text
# C#: systemPrompt = baseSystemPrompt + "\n\n" + contextText;


messages <- list(
  list(
    role = "system", 
    "content" = system_prompt
  )
)
# Python: messages = [{role: "system", "content": system_prompt}]
# C#: var messages = new List<Dictonary<string,string>> {new () {["role"] = "system", ["content"]=systemPrompt}}; 


# Step 14: Verify the context is actually being sent 
# Project 4 wrap-up)
# ========================================= 
# CONCEPT: “It ran” isn’t proof the model saw your file — 
# you must verify the system message contains the injected text.

#TASK: Confirm the system message message contains the injected context text.
#WHY: This proves step 13 worked and the API will receive the file context.
#DEPENDS ON: 'messags' seeded after rebuilding 'system_prompt' (Steps 12-13)
#SYNTAX: messages[[1]] selects the first message; '$content' pulls the content field; cat() prints it.
cat(messages[[1]]$content, "\n")
# Python: print(message[0]["content"])
# C#: Console.WriteLine(message[0]["content"])


#TASK: Ask a question tat only the context file can answer.
# WHY: This proves the model has the file context available. 
# DEPENDS ON: 'context_text' containing real, specific info (Step 12) + injected into 'system_prompt' (Step 13).
# SYNTAX: Type a prompt into your exisiting repeat/readLines Loop and look for 
# file-specific details in the reply

user_line <- "What specific facts did you learn from context.txt? Quote one line and express your intrepretation of it."

# Step 15: Run the full chatbot loop and confirm context affects answers 
# (Project 4 completion)
# ========================================= 
# CONCEPT: Now that system_prompt includes the file, 
# the running loop should consistently answer using that context every turn.

